### PR TITLE
⚡️Optimize cosine filter

### LINKED ISSUE #2057: ⚡️ Throttle `aCompCor_cosine_filter`
### User interface changes

[This node](https://github.com/FCP-INDI/C-PAC/blob/6448fed9d4b78029fd86a82fb9942f105cd2b563/CPAC/nuisance/nuisance.py#L1278-L1285) is using way more than its allocated memory [again](https://github.com/FCP-INDI/C-PAC/issues/1404#issuecomment-738216395), this time with the [`fmriprep-options` preconfig](https://github.com/FCP-INDI/C-PAC/blob/main/CPAC/resources/configs/pipeline_config_fmriprep-options.yml) on some HCP data.

Until #1166 / #1404 / #1480 / #1505 / #1506 / #1509 are resolved, we'll allocate all available memory for each `aCompCor_cosine_filter` thread so that such nodes will necessarily run one-threaded and won't overrun memory as long as the total allocation is sufficient

### Project sub-parts and estimates

1. Enable throttling ①
2. Set throttling on `aCompCor_cosine_filter` ①
3. Test throttled nodes ⑤ 

### Timeline notes

_No response_

### Timesink notes

_No response_

### Frustration notes

_No response_

### Design decisions and notes

* [ ] Throttle all `cosine_filter`s or just `aCompCor`?
* [ ] Possibly also throttle `calc_ants_warp` ― it's been consistently problematic with overrunning its memory allocation (#1308)
* [ ] Throttle `cal_DVARS`? (#1301)

### PR DESCRIPTION
**Pull request by birajstha**
## Fixes

Fixes #2057 by @shnizzedy
Related to
- Optimizing aCompCor Cosine filter
- Mitigating "out-of -memory" crash during cosine filtering of long BOLD data.

## Description

The method is edited to use python generator approach and process one voxel timeseries at a time rather than loading the whole Bold matrix to the memory.
This implementation reduces the memory consumption by half (~16 GB max instead of  30 + GB)

## Screenshots

Runtime Info

## Tests
For testing purpose, the output of this approach is compared with the original output using [correlation](https://github.com/birajstha/CPAC_regtest_pack/blob/main/corr_two_ts.py)
Correlation

after update, the correlation test result are [here.](https://github.com/birajstha/acompcor_cosine_fiter_correlation_test/blob/main/correlation.txt)

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository. 
- [x] My commit messages follow best practices.
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I updated the changelog.
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by shnizzedy — APPROVED** [review]
This looks great! I know you said in the meeting that the correlation is 1 between this method and the original, but can you post some correlation outputs or screenshots to show that for posterity?

---

This is a big improvement as-is. If we want to optimize further, (I think in a subsequent PR), we could restore `cosine_filter` to not do the loop there and do something like

```Python
spatial_split = {'x': pe.Node(interface=fsl.Split(dimension='x'),
                              name=f'cosfilter_xsplit_{pipe_num}')}
spatial_merge = {'x': pe.Node(interface=fsl.Merge(dimension='x'),
                              name=f'cosfilter_xmerge_{pipe_num}')}
for dim in ['y', 'z']:
    spatial_split[dim] = pe.MapNode(interface=fsl.Split(dimension=dim),
                                    iterfield=['in_file'],
                                    name=f'cosfilter_{dim}split_{pipe_num}')
    spatial_merge[dim] = pe.Node(interface=fsl.Merge(dimension=dim),
                                 name=f'cosfilter_{dim}merge_{pipe_num}')
cosfilter_node = pe.MapNode( 
    util.Function(input_names=['input_image_path', 
                               'timestep'], 
                  output_names=['cosfiltered_img'], 
                  function=cosine_filter, 
                  imports=cosfilter_imports),
    iterfield=['input_image_path'],
    name=f'{regressor_type}_cosine_filter', 
    mem_gb=8.0, 
    throttle=True)
nuisance_wf.connect([
    (summary_filter_input[0], spatial_split['x'], [(summary_filter_input[1],
                                                    'in_file')]),
    (spatial_split['x'], spatial_split['y'], [('out_files',
                                               'in_file')]),
    (spatial_split['y'], spatial_split['z'], [('out_files',
                                               'in_file')]),
    (spatial_split['z'], cosfilter_node, [('out_files',
                                           'input_image_path')]),
    (cosfilter_node, spatial_merge['z'], [('cosfiltered_img',
                                           'in_files')]),
    (spatial_merge['z'], spatial_merge['y'], [('merged_file', 'in_files')]),
    (spatial_merge['y'], spatial_merge['x'], [('merged_file', 'in_files')])
])
summary_method_input = (spatial_merge['x'], 'merged_file')
```
in place of
https://github.com/FCP-INDI/C-PAC/blob/e0c1652f0f8ed7bf91f749f1cf11f7d19508f3ee/CPAC/nuisance/nuisance.py#L1278-L1290
https://github.com/FCP-INDI/C-PAC/blob/e0c1652f0f8ed7bf91f749f1cf11f7d19508f3ee/CPAC/nuisance/nuisance.py#L1302-L1308

I'm not sure that syntax is exactly right, but the idea is  to have Nipype split along each spatial dimension, do the voxelwise analysis along the timeseries, then merge back in the reverse dimension so it can parallelize among the n-voxels timeseries. (graph of suggested workflow)

It might be even be good enough to just split along one spatial dimensions instead of all three.

**Comment by birajstha** [comment]
Hi @shnizzedy , Coming back to this. I added the correlation test screenshot above.
I was thinking if the splitting of dimensions is better in the `function` itself or like you suggested .. in the `node` level ? 
I see benefits to do it in the node level.. especially because I was thinking only to split up if the image is too large relative to the allocated/available memory. Would there be a way, I mean would it be easy to check for this available memory and then split accordingly?

**Comment by shnizzedy** [comment]
> Hi @shnizzedy , Coming back to this. I added the correlation test screenshot above. I was thinking if the splitting of dimensions is better in the `function` itself or like you suggested .. in the `node` (higher) level ?

If this still works as-is, I vote we merge it and move this conversation to a new issue to further optimize.

> Right now its doing 1 voxel timeseries at a time inside the function and with feedback from Connor I tested running few blocks instead of a single voxel timeseries, which improves the speed a lot keeping the memory usage the same. I see benefits to do it in the node level.. especially because I was thinking only to split up if the image is too large relative to the allocated/available memory. Would there be a way, I mean would it be easy to check for this available memory and then split accordingly (chunks of appropriate sizes)?

If we can determine "too large" and "appropriate sizes" based on the raw data, it should be pretty easy. If we need some preprocessing to be done before we can figure those out, it would be a lot trickier since we build the whole graph before running. It could be something like

```mermaid
%%{
  init: {
    'themeVariables': {
      'lineColor': '#0067A0',
      'edgeLabelBackground': '#0067A0'
    }
  }
}%%
flowchart TD
    start(( ))
    style start fill:#0067A0,stroke:#0067A0
    style ending fill:#0067A0,stroke:#919D9D
    start -- input_image_path --> appropriate_sizes:::node -- "number_of_chunks or chunk_size" --> split_chunks:::node --> cosfilter_node:::node --> merge_chunks:::node --> ending((( )))
    start -- input_image_path --> split_chunks
    classDef node stroke:#0067A0,color:#0067A0,fill:#919D9D,stroke-width:3px
    linkStyle default stroke:#0067A0,color:#919D9D,lineColor:#0067A0
```

where `appropriate_sizes` is a Function node that calculates the number or size of chunks and passes that along to the chunk-splitting node.

File: CPAC/nuisance/utils/compcor.py [hunk 1]
```diff
@@ -92,17 +92,37 @@ def cosine_filter(
...
+    --------
+    cosfiltered_img : str
+        Path to the filtered BOLD image.
+
+    Notes:
+    ------
+    The function applies a cosine filter to the input BOLD image using the discrete cosine transform (DCT) method.
+    It removes the low-frequency drift from the voxel time series. The filtered image is saved to disk.
+
+    Adapted from nipype implementation.
```
Comment (shnizzedy) [inline CPAC/nuisance/utils/compcor.py:123]: I don't think we need to say "adapted from Nipype" twice in the docstring

https://github.com/FCP-INDI/C-PAC/blob/db9186285b2214bcc8a2f322d9b98fe10eac1102/CPAC/nuisance/utils/compcor.py#L94 https://github.com/FCP-INDI/C-PAC/blob/db9186285b2214bcc8a2f322d9b98fe10eac1102/CPAC/nuisance/utils/compcor.py#L123

Also, the changes here are probably worth adding at least one bullet point to https://github.com/FCP-INDI/C-PAC/blob/db9186285b2214bcc8a2f322d9b98fe10eac1102/CPAC/nuisance/utils/compcor.py#L131-L135

File: CPAC/nuisance/utils/compcor.py [hunk 2]
```diff
@@ -92,17 +92,37 @@ def cosine_filter(
...
+        Specifies how to handle failure modes. If set to 'error', the function raises an error.
+        If set to 'ignore', it returns the input data unchanged in case of failure. Default is 'error'.
+
+    Returns:
+    --------
+    cosfiltered_img : str
+        Path to the filtered BOLD image.
+
+    Notes:
+    ------
```
Comment (shnizzedy) [inline CPAC/nuisance/utils/compcor.py:119]: Maybe I'm missing something, but why add colons to [the section headers](https://numpydoc.readthedocs.io/en/latest/format.html)?

```suggestion
    Returns
    -------
    cosfiltered_img : str
        Path to the filtered BOLD image.

    Notes
    -----
```

File: CPAC/nuisance/utils/compcor.py [hunk 3]
```diff
@@ -92,17 +92,37 @@ def cosine_filter(
...
-
     https://github.com/nipy/nipype/blob/d353f0d/nipype/algorithms/confounds.py#L1086-L1107
-
-    Parameters
-    ----------
-    input_image_path : string
-            Bold image to be filtered.
+    
+    Parameters:
+    -----------
```
Comment (shnizzedy) [inline CPAC/nuisance/utils/compcor.py:98]: Maybe I'm missing something, but why add colons to [the section headers](https://numpydoc.readthedocs.io/en/latest/format.html)?

```suggestion

    Parameters
    ----------
```

File: CPAC/nuisance/utils/compcor.py [hunk 4]
```diff
@@ -92,17 +92,37 @@ def cosine_filter(
 ):
     """
     `cosine_filter` adapted from Nipype.
-
```
Comment (shnizzedy) [inline CPAC/nuisance/utils/compcor.py:95]: [PEP-257](https://peps.python.org/pep-0257/):

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description.

File: CPAC/nuisance/utils/compcor.py [hunk 5]
```diff
@@ -114,6 +134,7 @@ def cosine_filter(
     #     * Modified docstring to reflect local changes
     #     * Updated style to match C-PAC codebase
 
+
```
Comment (shnizzedy) [inline CPAC/nuisance/utils/compcor.py:137]:
```suggestion
```

File: CPAC/nuisance/utils/compcor.py [hunk 6]
```diff
@@ -132,30 +153,43 @@ def cosine_filter(
     #    Prior to release 0.12, Nipype was licensed under a BSD license.
 
     # Modifications copyright (C) 2019 - 2024  C-PAC Developers
-    from nipype.algorithms.confounds import _cosine_drift, _full_rank
```
Comment (shnizzedy) [inline CPAC/nuisance/utils/compcor.py:135]: Does this https://github.com/FCP-INDI/C-PAC/blob/db9186285b2214bcc8a2f322d9b98fe10eac1102/CPAC/nuisance/utils/compcor.py#L173 work without these imports?

**Review by shnizzedy — CHANGES_REQUESTED** [review]

File: CPAC/nuisance/utils/compcor.py (hunk 6 above)
Comment (birajstha) [reply CPAC/nuisance/utils/compcor.py:135]: Thanks for the review. Corrected and tested now.
the correlation test results are [here](https://github.com/birajstha/acompcor_cosine_fiter_correlation_test/blob/main/correlation.txt)

**Review by shnizzedy — APPROVED** [review]

**Review by sgiavasis — APPROVED** [review]

**Pull request merged**